### PR TITLE
feat: add renderer.icons.diagnostic_placement

### LINKED ISSUE #1510: Add nvim-tree.diagnostics.placement config option
**Is your feature request related to a problem? Please describe.**
No.

**Describe the solution you'd like**
It would be nice to have an option in the config such as `nvim-tree.diagnostics.placement`, which would act similarly to `nvim-tree.renderer.icons.git_placement`, with possible values `left` and `right`. This would allow the user to place diagnostic icons at the left/right end of the nvim-tree window; currently, this defaults to the left end of the window and cannot be changed (AFAIK).

**Describe alternatives you've considered**
N/A.

**Additional context**
N/A.

### PR DESCRIPTION
**Pull request by baahrens**
Attempts to close #1510 

This is my first proper PR in this project, so please let me know if this is the wrong approach/direction or if I broke any styling guidelines.

**There's a few things I'm unsure about:**

1. Is it okay to reload the whole explorer on `DiagnosticChanged` or should we aim for a more finegrained approach?
2. Should this be `config.renderer.icons.diagnostic_placement` or `config.diagnostics.placement` (this was proposed in #1510)?
3. There was a lot of logging in the original `diagnostics.update` function. Should this be replicated somewhere in the new implementation? If yes, where?

**TODO:**

- [x] Add support for folder diagnostics
- [ ] documentation/help

### CODE REVIEW

**Comment by alex-courtis** [comment]
> 1. Is it okay to reload the whole explorer on `DiagnosticChanged` or should we aim for a more finegrained approach?

Unfortunately we cannot afford that expensive operation.

Diagnostic changes must be updated in the debounced "thread" `M.update`, affecting only diagnostics.

**Comment by alex-courtis** [comment]
> 2\. Should this be `config.renderer.icons.diagnostic_placement` or `config.diagnostics.placement`

`config.diagnostics.placement` is preferable, however inconsistent with modified/git.

`config.renderer.icons.diagnostic_placement` please.

**Comment by alex-courtis** [comment]
> There was a lot of logging in the original diagnostics.update function. Should this be replicated somewhere in the new implementation? If yes, where?

Please retain the logging and profiling, in the `M.update` debounced "thread".

**Review by alex-courtis — CHANGES_REQUESTED** [review]
This is looking great - thanks for your work.

**Comment by baahrens** [comment]
Thanks for your feedback!

I adjusted a few things:
- The tree is now redrawing instead of reloaded on `DiagnosticChanged`
- Diagnostic are now shown on folders, respecting `diagnostics.show_on_dirs` and `diagnostics.show_on_open_dirs` (can you check, if this is working correctly? I'm not quite sure)
- Added the logging again, although I'm not sure it makes sense like this, as the `update` function from before doesn't exist anymore, as diagnostic rendering is now part of the drawing process and doesn't happen in a debounced "thread"

Let me know if this is going in the right direction, I'll update the help file then.


`renderer.icons.diagnostic_placement == "after"`
Screenshot 2023-04-24 at 22 38 11
`renderer.icons.diagnostic_placement == "before"`
Screenshot 2023-04-24 at 22 38 59

**Review by alex-courtis — CHANGES_REQUESTED** [review]
Looking good except for sign column folders.

Test case `nvim-tree/dev/lua/nvim-tree/actions/finders/find-file.lua`

```lua
  view = {
    width = {
      min = 20,
      max = 50,
    },
  },
```

OK:
`show_on_dirs = false,` `show_on_open_dirs = false,`
`diagnostic_placement = "before",`
`diagnostic_placement = "signcolumn",`
`diagnostic_placement = "after",`
`diagnostic_placement = nil`,

`show_on_dirs = true,` `show_on_open_dirs = false,`
`diagnostic_placement = "before",`
`diagnostic_placement = "signcolumn",`
`diagnostic_placement = "after",`
`diagnostic_placement = nil`,

`show_on_dirs = false,` `show_on_open_dirs = true,`
`diagnostic_placement = "before",`
`diagnostic_placement = "signcolumn",`
`diagnostic_placement = "after",`
`diagnostic_placement = nil`,

`show_on_dirs = true,` `show_on_open_dirs = true,`
`diagnostic_placement = "before",`
`diagnostic_placement = "after",`

FAIL:
`show_on_dirs = true,` `show_on_open_dirs = true,`
`diagnostic_placement = "signcolumn",`
`diagnostic_placement = nil`,

Same results with `group_empty = true,` in an empty directory tree.

**Comment by alex-courtis** [comment]
> * 
> 
> * Added the logging again, although I'm not sure it makes sense like this, as the `update` function from before doesn't exist anymore, as diagnostic rendering is now part of the drawing process and doesn't happen in a debounced "thread"

Please wrap the profile / logging around the `"DiagnosticChanged"` and `"CocDiagnosticChange"` autocommands. We want to capture the time spent fetching diagnostics.

File: lua/nvim-tree/renderer/builder.lua [hunk 1]
```diff
@@ -364,13 +406,44 @@ end
 
 function Builder:build(tree, unloaded_bufnr)
   local num_children = self:_get_nodes_number(tree.nodes)
+  local diagnostic_severities_by_path = diagnostics.get_diagnostics()
```
Comment (alex-courtis) [inline lua/nvim-tree/renderer/builder.lua:409]: This doesn't fit in here; it should go back in diagnostics.lua.

I like `_get_diagnostic_icon` - nice and consistent.

Proposal:

- Move this back into `diagnostics.update`
- `diagnostics.update` records the diagnostic state of the node. Something like `git_status`
- call draw following `diagnostics.update`
- `_get_diagnostic_icon` calls something like `local diagnostic_icon = diagnostics.get_icon(node)`

**Comment by alex-courtis** [comment]
Any updates on this one @baahrens ? This would be a great addition... I will be using it...

**Comment by baahrens** [comment]
I will look into this shortly, sorry for the delay

**Comment by alex-courtis** [comment]
> I will look into this shortly, sorry for the delay

There is no hurry; just checking whether you still intend to work on this one.

**Comment by alex-courtis** [comment]
superseded by #2396

**Pull request closed without merging**